src/metatrain/experimental/nanopet/model.py

```diff
@@ -0,0 +1,474 @@
+                )
+                corresponding_new_features = new_features[corresponding_edges]
+                new_features = torch.concatenate(
+                    [new_features, corresponding_new_features], dim=-1
+                )
+                new_features = contraction(new_features)
+                new_features = edge_array_to_nef(new_features, nef_indices)
+                new_features = transformer(new_features, radial_mask)
+                features = (features + new_features) * 0.5**0.5
+
+        edge_features = features * radial_mask[:, :, None]
+        node_features = torch.sum(edge_features, dim=1)
+
+        return_dict: Dict[str, TensorMap] = {}
+
+        # output the hidden features, if requested:
+        if "mtt::aux::last_layer_features" in outputs:
+            last_layer_feature_tmap = TensorMap(
+                keys=self.single_label,
+                blocks=[
+                    TensorBlock(
+                        values=node_features,
+                        samples=sample_labels,
+                        components=[],
+                        properties=Labels(
+                            names=["properties"],
+                            values=torch.arange(
+                                node_features.shape[-1], device=node_features.device
+                            ).reshape(-1, 1),
+                        ),
+                    )
+                ],
+            )
+            last_layer_features_options = outputs["mtt::aux::last_layer_features"]
+            if last_layer_features_options.per_atom:
+                return_dict["mtt::aux::last_layer_features"] = last_layer_feature_tmap
+            else:
+                return_dict["mtt::aux::last_layer_features"] = (
+                    metatensor.torch.sum_over_samples(last_layer_feature_tmap, ["atom"])
+                )
+
+        atomic_properties_tmap_dict: Dict[str, TensorMap] = {}
+        for output_name, last_layer in self.last_layers.items():
+            if output_name in outputs:
+                atomic_properties = last_layer(node_features)
+                block = TensorBlock(
+                    values=atomic_properties.reshape(
+                        [-1] + self.output_shapes[output_name]
+                    ),
+                    samples=sample_labels,
+                    components=self.component_labels[output_name],
+                    properties=self.property_labels[output_name],
+                )
+                atomic_properties_tmap_dict[output_name] = TensorMap(
+                    keys=self.key_labels[output_name],
+                    blocks=[block],
+                )
+
+        if selected_atoms is not None:
+            for output_name, tmap in atomic_properties_tmap_dict.items():
+                atomic_properties_tmap_dict[output_name] = metatensor.torch.slice(
+                    tmap, axis="samples", selection=selected_atoms
+                )
```

Comment:
How hard would it be to skip the calculation completely if some centers are not part of the selected atoms? This might speed up calculation a bit in LAMMPS, and would help a lot with QM/ML simulations if we get there at some point.

Reply:
Very hard IMO... it might be easier, in case of MM/ML, to feed the region of interest to the model + a ghost region up to the interaction range (+ selected atoms)

Reply:
Ok, sucks a bit because this would make PET in LAMMPS a lot more efficient (right now it needs to compute some prediction for all atoms in the ghost region, and then throw it away).

Reply:
We can discuss in person